tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -201,6 +214,83 @@ private int Serialize(ref byte[] bytes, int offset, in SpanModel spanModel)
                 offset += WriteMetaStruct(ref bytes, offset, in spanModel);
             }
 
+            if (span.SpanLinks is { Count: > 0 })
+            {
+                offset += WriteSpanLink(ref bytes, offset, in spanModel);
+            }
+
+            return offset - originalOffset;
+        }
+
+        private int WriteSpanLink(ref byte[] bytes, int offset, in SpanModel spanModel)
+        {
+            int originalOffset = offset;
+            offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _spanLinkBytes);
+            offset += MessagePackBinary.WriteArrayHeader(ref bytes, offset, spanModel.Span.SpanLinks.Count);
+            foreach (var spanLink in spanModel.Span.SpanLinks)
+            {
+                var context = spanLink.Context;
+                var samplingPriority = context.TraceContext?.SamplingPriority ?? context.SamplingPriority;
+                var traceState = W3CTraceContextPropagator.CreateTraceStateHeader(context);
+                var traceFlags = samplingPriority switch
+                {
+                    null => 0u,             // not set
+                    > 0 => 1u + (1u << 31), // keep
+                    <= 0 => 1u << 31,       // drop
+                };
+                var len = 3;
+                // check to serialize tracestate
+                if (context.IsRemote)
+                {
+                    len++;
+                }
+
+                if (traceFlags > 0)
+                {
+                    len++;
+                }
+
+                if (spanLink.Attributes is { Count: > 0 })
+                {
+                    len++;
+                }
+
+                offset += MessagePackBinary.WriteMapHeader(ref bytes, offset, len);
+                // individual key-value pairs - traceid - lower
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _traceIdBytes);
+                offset += MessagePackBinary.WriteUInt64(ref bytes, offset, context.TraceId128.Lower);
+                // individual key-value pairs - traceid - higher
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _traceIdHighBytes);
+                offset += MessagePackBinary.WriteUInt64(ref bytes, offset, context.TraceId128.Upper);
+                // spanid
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _spanIdBytes);
+                offset += MessagePackBinary.WriteUInt64(ref bytes, offset, context.SpanId);
+                // optional serialization
+                // CreateTraceStateHeader will never return null or empty
+                if (context.IsRemote)
```

Comment:
We should move the line above that calls `W3CTraceContextPropagator.CreateTraceStateHeader(context)` inside of this `if`, so we don't call it unless we're going to add the `tracestate` to the span link. (avoiding unnecessary work)